File: tests/playwright/playwright_static_server.js
const http = require('http');
const fs = require('fs');
const path = require('path');
const { URL } = require('url');

const repoRoot = path.resolve(__dirname, '..', '..');
const staticRoot = path.resolve(repoRoot, 'static');

function parseArgs(argv) {
  const out = { host: '127.0.0.1', port: 4173 };
  for (let i = 2; i < argv.length; i += 1) {
    const key = String(argv[i] || '').trim();
    const value = String(argv[i + 1] || '').trim();
    if (key === '--host' && value) {
      out.host = value;
      i += 1;
      continue;
    }
    if (key === '--port' && value) {
      const n = Number(value);
      if (Number.isInteger(n) && n > 0 && n < 65536) out.port = n;
      i += 1;
    }
  }
  return out;
}

function contentType(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  switch (ext) {
    case '.html': return 'text/html; charset=utf-8';
    case '.js': return 'application/javascript; charset=utf-8';
    case '.css': return 'text/css; charset=utf-8';
    case '.json': return 'application/json; charset=utf-8';
    case '.svg': return 'image/svg+xml';
    case '.png': return 'image/png';
    case '.jpg':
    case '.jpeg': return 'image/jpeg';
    case '.gif': return 'image/gif';
    case '.woff': return 'font/woff';
    case '.woff2': return 'font/woff2';
    default: return 'application/octet-stream';
  }
}

function resolveStaticFile(requestUrl) {
  let pathname = '/';
  try {
    pathname = new URL(requestUrl, 'http://127.0.0.1').pathname;
  } catch (_) {
    return null;
  }
  if (!pathname.startsWith('/static/')) return null;
  const relative = pathname.slice('/static/'.length);
  if (!relative) return null;
  const candidate = path.resolve(staticRoot, relative);
  if (!candidate.startsWith(staticRoot + path.sep) && candidate !== staticRoot) return null;
  if (!fs.existsSync(candidate) || !fs.statSync(candidate).isFile()) return null;
  return candidate;
}

const args = parseArgs(process.argv);
const server = http.createServer((req, res) => {
  const method = String(req.method || '').toUpperCase();
  if (method !== 'GET' && method !== 'HEAD') {
    res.writeHead(405, { 'Content-Type': 'text/plain; charset=utf-8' });
    res.end('Only GET and HEAD are allowed.');
    return;
  }
  const filePath = resolveStaticFile(req.url || '/');
  if (!filePath) {
    res.writeHead(404, { 'Content-Type': 'text/plain; charset=utf-8' });
    res.end('Not found.');
    return;
  }
  fs.readFile(filePath, (error, body) => {
    if (error) {
      res.writeHead(500, { 'Content-Type': 'text/plain; charset=utf-8' });
      res.end('Failed to read file.');
      return;
    }
    res.writeHead(200, {
      'Content-Type': contentType(filePath),
      'Content-Length': String(body.length),
      'Cache-Control': 'no-store',
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': 'no-referrer',
    });
    if (method === 'HEAD') {
      res.end();
      return;
    }
    res.end(body);
  });
});

function shutdown() {
  server.close(() => {
    process.exit(0);
  });
  // Failsafe: force-exit if close callback does not fire.
  setTimeout(() => process.exit(0), 1000).unref();
}

process.on('SIGINT', shutdown);
process.on('SIGTERM', shutdown);
process.on('SIGHUP', shutdown);

server.listen(args.port, args.host, () => {
  // Keep the line stable because Playwright prints it in webServer logs.
  console.log(`Serving playwright static assets on http://${args.host}:${args.port}/static/home.html`);
});

File: tests/playwright/playwright_static_server.py
from __future__ import annotations

import argparse
import mimetypes
from http import HTTPStatus
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import unquote, urlsplit


REPO_ROOT = Path(__file__).resolve().parents[2]
STATIC_ROOT = REPO_ROOT / "static"


class StaticOnlyHandler(BaseHTTPRequestHandler):
    server_version = "zizai-playwright/1.0"
    protocol_version = "HTTP/1.1"

    def do_GET(self) -> None:  # noqa: N802
        self._serve()

    def do_HEAD(self) -> None:  # noqa: N802
        self._serve(send_body=False)

    def do_POST(self) -> None:  # noqa: N802
        self.send_error(HTTPStatus.METHOD_NOT_ALLOWED, "Only GET and HEAD are allowed.")

    def do_PUT(self) -> None:  # noqa: N802
        self.send_error(HTTPStatus.METHOD_NOT_ALLOWED, "Only GET and HEAD are allowed.")

    def do_DELETE(self) -> None:  # noqa: N802
        self.send_error(HTTPStatus.METHOD_NOT_ALLOWED, "Only GET and HEAD are allowed.")

    def list_directory(self, path):  # type: ignore[override]
        self.send_error(HTTPStatus.FORBIDDEN, "Directory listing is disabled.")
        return None

    def log_message(self, format: str, *args) -> None:  # noqa: A003
        return

    def _serve(self, send_body: bool = True) -> None:
        file_path = self._resolve_path()
        if file_path is None:
            self.send_error(HTTPStatus.NOT_FOUND, "Not found.")
            return

        try:
            content = file_path.read_bytes()
        except OSError:
            self.send_error(HTTPStatus.INTERNAL_SERVER_ERROR, "Failed to read file.")
            return

        content_type, _ = mimetypes.guess_type(str(file_path))
        body = content if send_body else b""
        self.send_response(HTTPStatus.OK)
        self.send_header("Content-Type", content_type or "application/octet-stream")
        self.send_header("Content-Length", str(len(content)))
        self.send_header("Cache-Control", "no-store")
        self.send_header("X-Content-Type-Options", "nosniff")
        self.send_header("Referrer-Policy", "no-referrer")
        self.end_headers()
        if send_body:
            self.wfile.write(body)

    def _resolve_path(self) -> Path | None:
        raw_path = urlsplit(self.path).path
        path = unquote(raw_path or "/")
        if not path.startswith("/static/"):
            return None
        relative = path[len("/static/") :]
        candidate = (STATIC_ROOT / relative).resolve()
        try:
            candidate.relative_to(STATIC_ROOT.resolve())
        except ValueError:
            return None
        if not candidate.is_file():
            return None
        return candidate


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=4173)
    args = parser.parse_args()

    httpd = ThreadingHTTPServer((args.host, args.port), StaticOnlyHandler)
    print(f"Serving playwright static assets on http://{args.host}:{args.port}/static/home.html", flush=True)
    try:
        httpd.serve_forever()
    finally:
        httpd.server_close()


if __name__ == "__main__":
    main()
